### PR TITLE
Add generic context option for custom function resolution

### PR DESCRIPTION
**Pull request by michaelbridge**
### Context
Enable pass-through of generic parameters to underlying custom functions via `context` option. Further discussion [here](https://github.com/handsontable/hyperformula/discussions/808).

### How did you test your changes?
Added tests; ran `npm run test`

### Types of changes
- [ ] Breaking change (a fix or a feature because of which an existing functionality doesn't work as expected anymore)
- [x] New feature or improvement (a non-breaking change that adds functionality)
- [ ] Bug fix (a non-breaking change that fixes an issue)
- [ ] Additional language file, or a change to an existing language file (translations)
- [ ] Change to the documentation

### Related issues:
1. Fixes #808

### Checklist:

- [x] I have reviewed the guidelines about [Contributing to HyperFormula](https://hyperformula.handsontable.com/guide/contributing.html) and I confirm that my code follows the code style of this project.
- [x] I have signed the [Contributor License Agreement](https://goo.gl/forms/yuutGuN0RjsikVpM2).
- [x] My change is compliant with the [OpenDocument](https://docs.oasis-open.org/office/OpenDocument/v1.3/os/part4-formula/OpenDocument-v1.3-os-part4-formula.html) standard.
- [x] My change is compatible with Microsoft Excel.
- [x] My change is compatible with Google Sheets.
- [x] I described my changes in the [CHANGELOG.md](https://github.com/handsontable/hyperformula/blob/master/CHANGELOG.md) file.
- [ ] My changes require a documentation update.
- [ ] My changes require a migration guide.

### CODE REVIEW

File: test/config.spec.ts [hunk 1]
```diff
@@ -97,6 +97,14 @@ describe('Config', () => {
     expect(() => new Config({caseFirst: 'abcd'})).toThrowError('Expected one of \'upper\' \'lower\' \'false\' for config parameter: caseFirst')
   })
 
+  it('validation: context', () => {
+    expect(() => new Config({context: 'ABCD'})).toThrowError('Expected value of type: object for config parameter: context')
```
Comment (sequba) [inline test/config.spec.ts:101]: Do you think we should validate the type of the `context` parameter? In my opinion, any type should be allowed.

File: test/config.spec.ts [hunk 2]
```diff
@@ -97,6 +97,14 @@ describe('Config', () => {
...
+    const context ={a: 1, b: 2}
+    const validConfig = {context}
+    const config = new Config(validConfig)
+    expect(config.context).toBe(context)
+  })
+
   it('should throw error when Function Translation cannot be found', () => {
     const config = new Config()
     const functionName = '0123456ABCDEFGH'
```
Comment (sequba) [inline test/config.spec.ts:110]: Please, add also a unit test that demonstrates the purpose of the `context` parameter. Create a custom function that uses the provided data. You may have a look at `test/custom-functions.spec.ts` to see how to create a custom function inside a test file.

File: src/Config.ts [hunk 3]
```diff
@@ -81,6 +81,14 @@ export interface ConfigParams {
    * @category Engine
    */
   chooseAddressMappingPolicy: ChooseAddressMapping,
+  /**
+   * A generic context object that can be used to pass data to custom functions.
+   *
+   * @default {}
```
Comment (sequba) [inline src/Config.ts:87]:
```suggestion
   * @default undefined
```

File: src/Config.ts [hunk 4]
```diff
@@ -494,6 +502,7 @@ export class Config implements ConfigParams, ParserConfig {
     currencySymbol: ['$'],
     caseSensitive: false,
     caseFirst: 'lower',
+    context: {},
```
Comment (sequba) [inline src/Config.ts:505]:
```suggestion
    context: undefined,
```

File: src/Config.ts [hunk 5]
```diff
@@ -596,6 +605,8 @@ export class Config implements ConfigParams, ParserConfig {
   public readonly currencySymbol: string[]
   /** @inheritDoc */
   public readonly undoLimit: number
+  /** @inheritDoc */
+  public readonly context: Record<string, unknown>
```
Comment (sequba) [inline src/Config.ts:609]:
```suggestion
  public readonly context: unknown
```

**Review by sequba — CHANGES_REQUESTED** [review]
Thank you for contributing to the HyperFormula project. Please look at my comments about the code.

**Review by sequba — APPROVED** [review]

**Review by sequba — APPROVED** [review]
Good job! After @evanSe approval, it can be merged to `develop` and released with the next version fo HyperFormula.

**Review by evanSe — APPROVED** [review]

**Comment by sequba** [comment]
Merging to temporary branch `feature/issue-1396` to resolve conflicts with `develop`

**Pull request merged**